scipy/sparse/_csparsetools.pyx.in

```diff
@@ -461,3 +461,30 @@ cdef _fill_dtype_map2(map):
 
 _fill_dtype_map(_LIL_FANCY_GET_DISPATCH, np.typecodes['Integer'])
 _fill_dtype_map2(_LIL_FANCY_SET_DISPATCH)
+
+
+def lil_matvec(int M, int N,
+               object[:] rows,
+               object[:] data,
+               cnp.ndarray vec,
+               cnp.ndarray result):
+    """
+    Compute y = A * x for LIL matrix A and dense vectors x and y.
+
+    Parameters
+    ----------
+    M, N, rows, data
+        The data from the LIL matrix A.
+    vec
+        The vector x as array.
+    res
+        The vector y as array.
+        Must be preallocated to shape `N`!
+
+    """
+
```

Comment:
Might be useful to `cdef` the local variables here. I'd take a look at the result of `cython -a` to see the impact.

Reply:
How can I do this with handling float as well as complex values?

Reply:
Okay, `cython.numeric` might be the solution.